regular_expressions/regular_expressions.md

```diff
@@ -0,0 +1,623 @@
+
+</div>
+
+
+The goal of regular expressions is to describe patterns based in which characters occur, and how many, in what order.  You can also put in a "capturing" element, to capture a specific part of the pattern (in our  example, extracting the patient identifier and/or the sample number).
+
+## A Great Tool to Start With
+
+Before we explain the details of how to construct regular expressions, it can be helpful to introduce the online regex checker as a great resource to learn and to check your regex before trying to use it in a script.
+
+We like "Regex 101", at https://www.regex101.com, but you can also search online for "regex checker" to find other, similar websites.  Regex 101 allows you to enter your regular expression as well as some sample text that you want to apply the regular expression to.  Below, or by [trying it out yourself](https://regex101.com/r/JJB8T3/1), you can see that we put in our "phone number" regular expression (at the top of the screen, in the appropriately labeled box) as well as the different varieties of phone number we want to make sure match our pattern (pasted in the larger text area below).  Matches are highlighted, and the number of matches is given (in our case, all five varieties of phone number do indeed match our pattern).  
+
+![Regex 101 application with sample phone number regex pattern and test phone number strings.  The right side has three helpful sections which are described below.](media/regex101.png)<!-- style = "border: 1px solid rgb(var(--color-highlight));" -->
+
+On the right side, there are helpful hints in three different scrollable windows.
+
+**Explanation Section**
+
+At the top of the right pane, each segment of the regular expression is parsed, or explained.  For example, we read the explanation of our first bit of regex as follows:
+
+> `\(` matches the character `(` with index 40<sub>10</sub> (28<sub>16</sub> or 50<sub>8</sub>) literally (case sensitive)<br/><br/>
+> ? matches the previous token between zero and one times, as many times as possible, giving back as needed (greedy)
+
+This explanation describes the first bit of regex.  First, it describes the character or characters indicated (in our case, the left parenthesis, `(`), then it describes the modifier that follows, which in our case is a question mark (`?`), which indicates that we want to find the indicated character 0-1 times in the pattern at that position.
+
+What does "index 40<sub>10</sub> (28<sub>16</sub> or 50<sub>8</sub>)" mean?  It's a description of the encoding of the `(` symbol in different number systems. The decimal code for left parenthesis is 40. This is equivalent to the hexadecimal (base 16 number that uses digits 0-F) number 28 and the octal (base 8 number that uses digits 0-7) 50.
```

Comment:
Is this something that could be in a highlight box that makes it clear you can use regex (or at least start using regex) without actually understanding these subscripted numbers?